src/HaroohiePals.Gui/ImGuiController.cs

```diff
@@ -160,390 +124,441 @@ void main()
-            UpdateImGuiInput(wnd);
+        conf.PixelSnapH = true;
+        conf.OversampleH = 1;
+        conf.OversampleV = 1;
 
-            _frameBegun = true;
-            ImGui.NewFrame();
-            //ImGuizmo.BeginFrame();
+        if (isIconFont)
+        {
+            conf.MergeMode = true;
+            conf.GlyphOffset.Y = 1;
+            conf.GlyphMinAdvanceX = 16f; // Use if you want to make the icon monospaced
         }
 
-        /// <summary>
-        /// Sets per-frame data based on the associated window.
-        /// This is called by Update(float).
-        /// </summary>
-        private void SetPerFrameImGuiData(float deltaSeconds)
+        return conf;
+    }
+
+    private unsafe void AddFont(ImGuiFont font, in ImFontConfigPtr fontConfig, ushort[] iconRanges = null)
+    {
+        var io = ImGui.GetIO();
+
+        fixed (byte* p = font.Data)
         {
-            ImGuiIOPtr io = ImGui.GetIO();
-            io.DisplaySize = new System.Numerics.Vector2(
-                _windowWidth / _scaleFactor.X,
-                _windowHeight / _scaleFactor.Y);
-            io.DisplayFramebufferScale = _scaleFactor;
-            io.DeltaTime = deltaSeconds; // DeltaTime is in seconds.
+            fixed (ushort* rang = iconRanges)
+                io.Fonts.AddFontFromMemoryTTF((IntPtr)p, font.Data.Length, font.Size * _uiScale, fontConfig, (IntPtr)rang);
         }
+    }
 
-        private void UpdateImGuiInput(GameWindow wnd)
-        {
-            var io = ImGui.GetIO();
+    /// <summary>
+    /// Recreates the device texture used to render text.
+    /// </summary>
+    private void RecreateFontDeviceTexture()
+    {
+        var io = ImGui.GetIO();
 
-            var mouseState = wnd.MouseState;
-            var keyboardState = wnd.KeyboardState;
+        Dictionary<ImGuiIconGlyph, int> registeredIcons = new();
 
-            io.AddMousePosEvent(mouseState.X, mouseState.Y);
-            io.AddMouseButtonEvent(0, mouseState.IsButtonDown(MouseButton.Left));
-            io.AddMouseButtonEvent(1, mouseState.IsButtonDown(MouseButton.Right));
-            io.AddMouseButtonEvent(2, mouseState.IsButtonDown(MouseButton.Middle));
-            io.AddMouseButtonEvent(3, mouseState.IsButtonDown(MouseButton.Button1));
-            io.AddMouseButtonEvent(4, mouseState.IsButtonDown(MouseButton.Button2));
-            io.AddMouseWheelEvent(mouseState.ScrollDelta.X, mouseState.ScrollDelta.Y);
+        foreach (var icon in _icons)
+            registeredIcons.Add(icon, RegisterIcon(icon.CodePoint, 16));
 
-            foreach (var key in KeyValues)
-            {
-                if (key == Keys.Unknown)
-                {
-                    continue;
-                }
+        io.Fonts.GetTexDataAsRGBA32(out IntPtr pixels, out int width, out int height, out int bytesPerPixel);
 
-                if (_imguiKeyMap.TryGetValue(key, out ImGuiKey imguiKey))
+        void loadIcon(int rectId, byte[] iconBytes)
```

Comment:
This local function is useless as it's only used in one place. You can either just put it as foreach body, or move it completely out of the method and add the extra necessary arguments. I think moving out is the best as it keeps the methods small.